File: week7/dataprocessing.go
// Data Processing System, Go.
package main

import (
	"fmt"
	"log"
	"math/rand"
	"os"
	"sync"
	"time"
	"runtime"
	"strconv"
	"strings"
)

// Task structure
type Task struct {
	ID       int
	Data	 string
}

// Result structure
type Result struct {
	TaskID       int
	ResultData   string
}

// Processing time
func (task *Task) process() string {
	// Simulate processing time
	delay := time.Duration(rand.Intn(100)) * time.Millisecond
	time.Sleep(delay)
	return fmt.Sprintf("Processed data for task %d: %s" , task.ID, task.Data)
	}

// Worker tasks
func worker(tasks <-chan Task, results chan<- Result, wg *sync.WaitGroup) {
	defer wg.Done()
	log.Printf("Worker %s started", getgoRoutineID())
	for task := range tasks {
		processingResult := task.process()
		result := Result{TaskID: task.ID, ResultData: processingResult}
		results <- result
		log.Printf("Worker %s processed task %d", getgoRoutineID(), task.ID)
	}
	log.Printf("Worker %s finished", getgoRoutineID())
}

// Get the ID of the current routine
func getgoRoutineID() uint64 {
	var buf [64]byte
	n := runtime.Stack(buf[:], false)
	idField := strings.Fields(strings.TrimPrefix(string(buf[:n]), "goroutine"))[0]
	id, err := strconv.ParseUint(idField, 10, 64)
	// Error handling
	if err != nil {
		panic(fmt.Sprintf("Failed to get goroutine ID: %v", err))
	}
	return id
}

// Main function
func main() {
	numWorkers := 10
	numTasks := 40
	tasks := make(chan Task, numTasks)
	results := make(chan Result, numTasks)
	var wg sync.WaitGroup


	// Start worker routines
	for i := 0; i < numWorkers; i++ {
		wg.Add(1)
		go worker(tasks, results, &wg)
	}

	// Generate tasks
	for i := 0; i < numTasks; i++ {
		task := Task{ID: i, Data: fmt.Sprintf("Data %d", i)}
		tasks <- task
		log.Printf("Main routine created task %d", i)
	}
	// Close the tasks channel, signaling no more tasks will be sent.
	close(tasks)

	// Wait for all workers to finish/ Collect all results
	go func() {
		wg.Wait()
		close(results)
	}()

	// Collect results and save to file
	outputFile := "results_go.txt"
	file, err := os.Create(outputFile)
	// Error handling, Check if the file was created successfully
	if err != nil {
		log.Fatalf("Failed to create output file: %v", err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalf("Failed to close output file: %v", err)
		}
	}()

	log.Printf("Main routine started collecting results to file...")
	for result := range results {
		_, err := file.WriteString(fmt.Sprintf("%+v\n", result))
		// Error handling, Check if the result was written successfully.
		if err != nil {
			log.Fatalf("Failed to write result to file: %v", err)
		}
	}
	// Confirming printing results to output file.
	log.Printf("Results have been written to %s", outputFile)
	log.Printf("Main routine finished")
}

File: week7/dataprocessing.java
package week7;

// Data Sharing System, Java.
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.*;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.logging.Level;
import java.util.logging.Logger;

// Task Class structure.
class Task {
    private final int id;
    private final String data;

    public Task(int id, String data){
        this.id = id;
        this.data = data;
    }

    public int getId() {
        return id;
    }

    public String getData() {
        return data;
    }

    // Processing time.
    public String process(){
        try {
            //Simulate processing time
            Thread.sleep(100);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        return "Processed data for task " + id + ": " + data;
    }
}

// Result Class structure.
class Result {
    private final int taskId;
    private final String resultData;

    public Result(int taskId, String resultData){
        this.taskId = taskId;
        this.resultData = resultData;
    }
    @Override
    public String toString() {
        return "Result [" +
                "Task ID =" + taskId +
                ", Data ='" + resultData +
                "]";
    }
}

// Worker class with task processing.
class Worker implements Runnable{
    private static final Logger logger = Logger.getLogger(Worker.class.getName());
    private final BlockingQueue<Task> tasks;
    private final List<Result> results;
    private final Lock resultsLock;

    public Worker(BlockingQueue<Task> tasks, List<Result> results, Lock resultsLock){
        this.tasks = tasks;
        this.results = results;
        this.resultsLock = resultsLock;
    }
    @Override
    public void run() {
        logger.log(Level.INFO, "{0} started processing tasks.", Thread.currentThread().getName());
        // Process tasks until interrupted or exit signal is received.
        try {
            while (true) {
                // Poll with timeout to allow for exiting.
                Task task = tasks.poll(1, TimeUnit.SECONDS);
                if (task == null) {
                    logger.log(Level.INFO, "{0} found no tasks to process. Exiting.", Thread.currentThread().getName());
                    break; // Exit if no task is available.
                }
                try {
                    String processingResult = task.process();
                    Result result = new Result(task.getId(), processingResult);
                    // Locking to ensure thread-safe access to results.
                    resultsLock.lock();
                    try {
                        results.add(result);
                        logger.log(Level.INFO, "{0} processed task {1}.", new Object[]{Thread.currentThread().getName(), task.getId()});
                    } finally {
                        resultsLock.unlock();
                    }
                // Exception handling for task processing.
                } catch (Exception e) {
                    logger.log(Level.SEVERE, "{0} encountered an error processing task " + task.getId() + "from " + Thread.currentThread().getName(), e);               
                }
            }
        } catch (InterruptedException e) {
            logger.log(Level.WARNING, "{0} was interrupted.", Thread.currentThread().getName());
            Thread.currentThread().interrupt();
        }
    }
}

// Main class for data sharing system.
public class dataprocessing {
    private static final Logger logger = Logger.getLogger(dataprocessing.class.getName());
    private static final int numWorkers = 10;
    private static final int numTasks = 40;
    private static final String outputFile = "results_java.txt";

    public static void main (String[] args) {
        BlockingQueue<Task> tasks = new LinkedBlockingQueue<>();
        List<Result> results = new ArrayList<>();
        Lock resultsLock = new ReentrantLock();

        // Create and start worker threads.
        ExecutorService executor = Executors.newFixedThreadPool(numWorkers);
        for (int i = 0; i < numWorkers; i++) {
            executor.execute(new Worker(tasks, results, resultsLock));
        }

        // Generate and add tasks to the queue.
        for (int i = 0; i < numTasks; i++) {
            Task task = new Task(i, "Data " + i);
            tasks.offer(task);
            logger.log(Level.INFO, "Task {0} added to the tasks list.", task.getId());
        }

        // Shutdown executor service.
        executor.shutdown();
        try {
            if (!executor.awaitTermination(10, TimeUnit.SECONDS)) {
                executor.shutdownNow();
            }
        } catch (InterruptedException e) {
            logger.log(Level.SEVERE, "Executor service interrupted during shutdown.", e);
            Thread.currentThread().interrupt();
        }

        // Writing and saving results to file.
        try (FileWriter writer = new FileWriter(outputFile)) {
            resultsLock.lock();
            try{
                for (Result result : results) {
                writer.write(result.toString() + "\n");
                }
            } finally {
                resultsLock.unlock();
            }
            //Confirming results written to file.
            logger.log(Level.INFO, "Results written to {0}.", outputFile);
        } catch (IOException e) {
            logger.log(Level.SEVERE, "Error writing results to file.", e);
        }
        // Confirming tasks completed and results saved.
        logger.log(Level.INFO, "All tasks processed and results saved.");
    }
}
